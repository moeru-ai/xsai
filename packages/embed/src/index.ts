export { embed as default, embed, type EmbedOptions, type EmbedResponseUsage, type EmbedResult } from './utils/embed'
export { embedMany, type EmbedManyOptions, type EmbedManyResult } from './utils/embed-many'
