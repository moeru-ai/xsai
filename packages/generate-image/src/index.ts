import type { CommonRequestOptions } from '@xsai/shared'

import { requestBody, requestHeaders, requestURL, responseCatch, responseJSON } from '@xsai/shared'

export interface GenerateImageOptions extends CommonRequestOptions {
  [key: string]: unknown
  /**
   * The number of images to generate.
   * @default `1`
   */
  n?: number
  /** A text description of the desired image(s). */
  prompt: string
  /**
   * The format in which generated images with `dall-e-2` and `dall-e-3` are returned from the API.
   * `gpt-image-1` only supports `b64_json`, and `url` will be ignored.
   *
   * Images will always be Base64-encoded in the result. If `url` is used, the images will be
   * fetched upon receiving the response.
   */
  responseFormat?: 'b64_json' | 'url'
  /**
   * The size of the generated images.
   * @default `1024x1024`
   */
  size?: `${number}x${number}`
}

export interface GenerateImageResponse {
  created: number
  data: {
    /**
     * The Bse64-encoded JSON of the generated image. Default value for `gpt-image-1`, and only
     * present if `response_format` is set to `b64_json` for `dall-e-2` and `dall-e-3`.
     */
    b64_json?: string
    /**
     * For `dall-e-3` only, the revised prompt that was used to generate the image.
     */
    revised_prompt?: string
    /**
     * When using `dall-e-2` or `dall-e-3`, the URL of the generated image if `response_format` is
     * set to `url` (default value). Unsupported for `gpt-image-1`.
     */
    url?: string
  }[]
}

export interface GenerateImageResult {
  image: GenerateImageResultImage
  images: GenerateImageResultImage[]
}

export interface GenerateImageResultImage {
  base64: string
  mimeType: string
}

/** @internal */
const mimeTypes = {
  '/9j/': 'image/jpg',
  'AAAAIGZ0eXBhdmlm': 'image/avif',
  'iVBORw0KGgo': 'image/png',
  'R0lGOD': 'image/gif',
  'UklGRg==': 'image/webp',
}

/** @internal */
const convertImage = (b64_json: string) => {
  const key = Object.keys(mimeTypes).find(prefix => b64_json.startsWith(prefix)) as keyof typeof mimeTypes | undefined
  const mimeType = mimeTypes[key ?? 'iVBORw0KGgo']

  return {
    // eslint-disable-next-line @masknet/string-no-data-url
    base64: `data:${mimeType};base64,${b64_json}`,
    mimeType,
  }
}

/** @internal */
const responseBlobAsDataURL = async (res: Response): Promise<string> =>
  responseCatch(res)
    .then(async (res) => {
      const blob = await res.blob()

      try {
        return await new Promise<string>((resolve, reject) => {
          const reader = new FileReader()
          reader.onloadend = () => resolve(reader.result as string)
          reader.onerror = reject
          reader.readAsDataURL(blob)
        })
      }
      catch {
        throw new Error(`Failed to parse response blob, response URL: ${res.url}`)
      }
    })

/** @experimental */
export const generateImage = async (options: GenerateImageOptions): Promise<GenerateImageResult> =>
  (options.fetch ?? globalThis.fetch)(requestURL('images/generations', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseJSON<GenerateImageResponse>)
    .then(async ({ data }) =>
      Promise.all(
        data.map(async (img, i) => {
          if (typeof img.b64_json === 'string') {
            return convertImage(img.b64_json)
          }

          if (typeof img.url === 'string') {
            return (options.fetch ?? globalThis.fetch)(new URL(img.url), {
              signal: options.abortSignal,
            })
              .then(responseBlobAsDataURL)
              .then((dataURL) => {
                const sepIndex = dataURL.indexOf(';')
                const mimeType = dataURL.substring(5, sepIndex) // `data:(...);...`
                return {
                  base64: dataURL,
                  mimeType,
                }
              })
          }

          throw new Error(`Unrecognized image at index ${i}: ${JSON.stringify(img)}`)
        }),
      ),
    )
    .then(images => ({ image: images[0], images }))
