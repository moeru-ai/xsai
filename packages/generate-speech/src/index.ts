import { type CommonRequestOptions, requestBody, requestHeaders, requestURL, responseCatch } from '@xsai/shared'

export interface GenerateSpeechOptions extends CommonRequestOptions {
  [key: string]: unknown
  input: string
  /** @default `mp3` */
  responseFormat?: 'aac' | 'flac' | 'mp3' | 'opus' | 'pcm' | 'wav'
  /** @default `1.0` */
  speed?: number
  voice: string
}

/**
 * @experimental WIP
 * @param options generate speech options.
 * @returns audio array buffer
 */
export const generateSpeech = async (options: GenerateSpeechOptions): Promise<ArrayBuffer> =>
  await (options.fetch ?? globalThis.fetch)(requestURL('audio/speech', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseCatch)
    .then(res => res.arrayBuffer())

export default generateSpeech
