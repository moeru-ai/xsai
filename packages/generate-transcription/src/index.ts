import type { CommonRequestOptions } from '@xsai/shared'

import { requestHeaders, requestURL, responseJSON } from '@xsai/shared'

export interface GenerateTranscriptionOptions extends CommonRequestOptions {
  file: Blob
  fileName?: string
  language?: string
  prompt?: string
  temperature?: string

  // response_format?: 'json' | 'text' | 'srt' | 'verbose_json' | 'vtt'
  // timestamp_granularities?: ('segment' | 'word')[]
}

export interface GenerateTranscriptionResult {
  text: string
}

export const generateTranscription = async (options: GenerateTranscriptionOptions): Promise<GenerateTranscriptionResult> => {
  const body = new FormData()

  body.append('model', options.model)
  body.append('file', options.file, options.fileName)
  options.language != null && body.append('language', options.language)
  options.prompt != null && body.append('prompt', options.prompt)
  options.temperature != null && body.append('temperature', options.temperature)

  return (options.fetch ?? globalThis.fetch)(requestURL('audio/transcriptions', options.baseURL), {
    body,
    headers: requestHeaders(options.headers, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseJSON<GenerateTranscriptionResult>)
    .then(({ text }) => ({ text: text.trim() }))
}
