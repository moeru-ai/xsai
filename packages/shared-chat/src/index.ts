export * from './error'
export type * from './types'
export * from './utils'
