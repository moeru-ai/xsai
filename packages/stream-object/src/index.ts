import type { StreamTextOptions, StreamTextResult } from '@xsai/stream-text'
import type { PartialDeep } from 'type-fest'
import type { Infer, Schema } from 'xsschema'

import { streamText } from '@xsai/stream-text'
import { parse } from 'best-effort-json-parser'
import { toJSONSchema } from 'xsschema'

export interface StreamObjectOptions<T extends Schema> extends StreamTextOptions {
  schema: T
  schemaDescription?: string
  schemaName?: string
}

export interface StreamObjectResult<T extends Schema> extends StreamTextResult {
  partialObjectStream: ReadableStream<PartialDeep<Infer<T>>>
}

/** @experimental WIP */
export const streamObject = async <T extends Schema>(options: StreamObjectOptions<T>): Promise<StreamObjectResult<T>> =>
  streamText({
    ...options,
    response_format: {
      json_schema: {
        description: options.schemaDescription,
        name: options.schemaName ?? 'json_schema',
        schema: await toJSONSchema(options.schema),
        strict: true,
      },
      type: 'json_schema',
    },
    schema: undefined,
    schemaDescription: undefined,
    schemaName: undefined,
  }).then(({ chunkStream, finishReason, textStream: rawTextStream, usage }) => {
    const [textStream, rawPartialObjectStream] = rawTextStream.tee()

    let partialObjectData = ''
    let partialObjectSnapshot = {} as PartialDeep<Infer<T>>
    const partialObjectStream = rawPartialObjectStream.pipeThrough(new TransformStream({
      transform: (chunk, controller) => {
        partialObjectData += chunk
        try {
          const data = parse(partialObjectData) as PartialDeep<Infer<T>>

          if (JSON.stringify(partialObjectSnapshot) !== JSON.stringify(data)) {
            partialObjectSnapshot = data
            controller.enqueue(data)
          }
        }
        catch {}
      },
    }))

    return {
      chunkStream,
      finishReason,
      partialObjectStream,
      textStream,
      usage,
    }
  })

export default streamObject
