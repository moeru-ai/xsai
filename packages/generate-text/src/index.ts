import type { AssistantMessageResponse, ChatOptions, FinishReason, Message, Tool, Usage } from '@xsai/shared-chat'

import { chat } from '@xsai/shared-chat'

export interface GenerateTextOptions extends ChatOptions {
  /** @default 1 */
  maxSteps?: number
  onStepFinish?: (step: GenerateTextStepResult) => Promise<void> | void
  /** @internal */
  steps?: GenerateTextStepResult[]
  /** if you want to enable stream, use `@xsai/stream-{text,object}` */
  stream?: never
  tools?: Tool[]
}

export interface GenerateTextResponse {
  choices: {
    finish_reason: FinishReason
    index: number
    message: AssistantMessageResponse
  }[]
  created: number
  id: string
  model: string
  object: 'chat.completion'
  system_fingerprint: string
  usage: Usage
}

export interface GenerateTextResult {
  finishReason: FinishReason
  messages: Message[]
  steps: GenerateTextStepResult[]
  text?: string
  toolCalls: GenerateTextToolCall[]
  toolResults: GenerateTextToolResult[]
  usage: Usage
}

export interface GenerateTextStepResult {
  finishReason: FinishReason
  stepType: 'continue' | 'initial' | 'tool-result'
  text?: string
  toolCalls: GenerateTextToolCall[]
  toolResults: GenerateTextToolResult[]
  usage: Usage
}

export interface GenerateTextToolCall {
  args: string
  toolCallId: string
  toolCallType: 'function'
  toolName: string
}

export interface GenerateTextToolResult {
  args: Record<string, unknown>
  result: string
  toolCallId: string
  toolName: string
}

/** @internal */
type RawGenerateText = (options: GenerateTextOptions) => RawGenerateTextTrampoline<GenerateTextResult>

/** @internal */
type RawGenerateTextTrampoline<T> = Promise<(() => RawGenerateTextTrampoline<T>) | T>

/** @internal */
const rawGenerateText: RawGenerateText = async (options: GenerateTextOptions) =>
  chat({
    ...options,
    maxSteps: undefined,
    messages: options.messages,
    steps: undefined,
    stream: false,
  })
    .then(async res => res.json() as Promise<GenerateTextResponse>)
    .then(async ({ choices, usage }) => {
      const messages: Message[] = structuredClone(options.messages)
      const steps: GenerateTextStepResult[] = options.steps ? structuredClone(options.steps) : []
      const toolCalls: GenerateTextToolCall[] = []
      const toolResults: GenerateTextToolResult[] = []

      const { finish_reason: finishReason, message } = choices[0]

      messages.push({ ...message, content: message.content! })

      const stepType = steps.length === 0
        ? 'initial'
        // eslint-disable-next-line sonarjs/no-nested-conditional
        : steps.at(-1)?.finishReason === 'tool-calls'
          ? 'tool-result'
          : 'continue'

      if ((message.content !== undefined && message.content.length > 0) || !message.tool_calls || steps.length >= (options.maxSteps ?? 1)) {
        const step: GenerateTextStepResult = {
          finishReason,
          stepType,
          text: message.content,
          toolCalls,
          toolResults,
          usage,
        }

        steps.push(step)

        if (options.onStepFinish)
          await options.onStepFinish(step)

        return {
          finishReason,
          messages,
          steps,
          text: message.content,
          toolCalls,
          toolResults,
          usage,
        }
      }

      for (const {
        function: { arguments: toolArgs, name: toolName },
        id: toolCallId,
        type: toolCallType,
      } of message.tool_calls) {
        const tool = options.tools?.find(tool => tool.function.name === toolName)

        if (!tool) {
          const availableTools = options.tools?.map(tool => tool.function.name)
          const availableToolsErrorMsg = availableTools === undefined
            ? 'No tools are available.'
            : `Available tools: ${availableTools.join(', ')}.`
          throw new Error(`Model tried to call unavailable tool '${toolName}. ${availableToolsErrorMsg}.`)
        }

        const parsedArgs = JSON.parse(toolArgs) as Record<string, unknown>
        const result = await tool.execute(parsedArgs, { abortSignal: options.abortSignal, messages, toolCallId })

        toolCalls.push({
          args: toolArgs,
          toolCallId,
          toolCallType,
          toolName,
        })

        toolResults.push({
          args: parsedArgs,
          result,
          toolCallId,
          toolName,
        })

        messages.push({
          content: result,
          role: 'tool',
          tool_call_id: toolCallId,
        })
      }

      const step: GenerateTextStepResult = {
        finishReason,
        stepType,
        text: message.content,
        toolCalls,
        toolResults,
        usage,
      }

      steps.push(step)

      if (options.onStepFinish)
        await options.onStepFinish(step)

      return async () => rawGenerateText({
        ...options,
        messages,
        steps,
      })
    })

export const generateText = async (options: GenerateTextOptions): Promise<GenerateTextResult> => {
  let result = await rawGenerateText(options)

  while (typeof result === 'function')
    result = await result()

  return result
}
