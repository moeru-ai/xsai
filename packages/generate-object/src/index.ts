import { type Infer, type Schema, toJSONSchema, validate } from '@typeschema/main'
import { generateText, type GenerateTextOptions, type GenerateTextResult } from '@xsai/generate-text'
import { clean } from '@xsai/shared'

export interface GenerateObjectOptions<T extends Schema> extends GenerateTextOptions {
  schema: T
  schemaDescription?: string
  schemaName?: string
}

// TODO: toolCalls, toolResults
export interface GenerateObjectResult<T extends Schema> extends Omit<GenerateTextResult, 'text' | 'toolCalls' | 'toolResults'> {
  object: Infer<T>
}

/** @experimental WIP */
export const generateObject = async <T extends Schema>(options: GenerateObjectOptions<T>): Promise<GenerateObjectResult<T>> =>
  await generateText({
    ...options,
    response_format: {
      json_schema: {
        description: options.schemaDescription,
        name: options.schemaName ?? 'json_schema',
        schema: await toJSONSchema(options.schema)
          .then(json => clean({
            ...json,
            $schema: undefined,
          })),
        strict: true,
      },
      type: 'json_schema',
    },
    schema: undefined,
    schemaDescription: undefined,
    schemaName: undefined,
  })
    .then(async ({ finishReason, steps, text, usage }) => {
      const result = await validate(options.schema, JSON.parse(text!))

      if (result.success) {
        return {
          finishReason,
          object: result.data,
          steps,
          usage,
        }
      }
      else {
        throw new Error([
          'Schema validation failed:',
          ...result.issues.map(issue => `- ${issue.message}`),
        ].join('\n'))
      }
    })

export default generateObject
