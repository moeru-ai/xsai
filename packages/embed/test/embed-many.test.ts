import { describe, expect, it } from 'vitest'

import { embedMany } from '../src'

describe('@xsai/embed', () => {
  const toFixed = (embedding: number[]) => embedding.map(e => e.toFixed(3))

  it('embedMany', async () => {
    const { embeddings, usage } = await embedMany({
      baseURL: 'http://localhost:11434/v1/',
      input: ['why is the sky blue?', 'why is the grass green?'],
      model: 'all-minilm',
    })

    expect(embeddings.map(toFixed)).toMatchSnapshot()
    expect(usage.prompt_tokens).toBe(12)
    expect(usage.total_tokens).toBe(12)
  })
})
