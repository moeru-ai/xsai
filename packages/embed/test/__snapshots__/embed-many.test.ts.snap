// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@xsai/embed > embedMany 1`] = `
[
  [
    "0.010",
    "-0.002",
    "0.050",
    "0.047",
    "0.055",
    "0.009",
    "0.105",
    "-0.026",
    "0.130",
    "0.032",
    "-0.044",
    "-0.009",
    "-0.001",
    "-0.064",
    "-0.016",
    "0.047",
    "-0.022",
    "-0.158",
    "-0.073",
    "-0.061",
    "-0.066",
    "0.054",
    "-0.062",
    "0.039",
    "-0.046",
    "0.055",
    "-0.035",
    "0.013",
    "0.043",
    "-0.008",
    "-0.019",
    "0.061",
    "0.037",
    "0.013",
    "-0.026",
    "-0.043",
    "0.073",
    "-0.049",
    "0.004",
    "-0.029",
    "-0.029",
    "-0.033",
    "-0.018",
    "0.016",
    "-0.012",
    "0.015",
    "-0.009",
    "0.026",
    "0.095",
    "-0.016",
    "-0.025",
    "0.009",
    "-0.077",
    "0.016",
    "0.050",
    "0.116",
    "0.001",
    "-0.020",
    "0.092",
    "0.008",
    "-0.057",
    "0.069",
    "-0.077",
    "0.069",
    "0.092",
    "-0.055",
    "-0.054",
    "0.008",
    "-0.063",
    "-0.066",
    "-0.025",
    "0.019",
    "0.061",
    "-0.028",
    "0.036",
    "0.001",
    "0.061",
    "-0.068",
    "-0.008",
    "-0.013",
    "0.031",
    "-0.064",
    "-0.074",
    "0.119",
    "0.013",
    "0.065",
    "0.015",
    "0.051",
    "-0.085",
    "0.010",
    "-0.008",
    "-0.036",
    "-0.115",
    "-0.031",
    "-0.083",
    "0.014",
    "0.057",
    "-0.041",
    "0.043",
    "0.022",
    "0.047",
    "-0.051",
    "0.030",
    "0.007",
    "-0.004",
    "-0.031",
    "0.078",
    "0.034",
    "0.061",
    "0.007",
    "-0.036",
    "-0.085",
    "0.022",
    "-0.019",
    "-0.040",
    "0.055",
    "-0.034",
    "0.018",
    "-0.106",
    "-0.050",
    "-0.012",
    "0.038",
    "0.022",
    "0.081",
    "0.008",
    "-0.017",
    "-0.059",
    "-0.000",
    "0.135",
    "-0.011",
    "0.092",
    "0.036",
    "0.040",
    "-0.055",
    "-0.035",
    "-0.004",
    "-0.020",
    "-0.035",
    "-0.006",
    "-0.015",
    "-0.024",
    "-0.048",
    "0.048",
    "-0.017",
    "-0.061",
    "0.006",
    "-0.083",
    "0.084",
    "-0.105",
    "0.042",
    "-0.037",
    "-0.008",
    "-0.028",
    "-0.043",
    "0.036",
    "0.075",
    "0.056",
    "0.012",
    "0.098",
    "0.105",
    "-0.022",
    "0.046",
    "-0.026",
    "-0.051",
    "-0.015",
    "-0.006",
    "-0.086",
    "0.029",
    "-0.054",
    "0.057",
    "-0.060",
    "0.012",
    "0.066",
    "-0.013",
    "0.038",
    "-0.089",
    "-0.057",
    "0.032",
    "-0.034",
    "0.024",
    "0.014",
    "-0.042",
    "0.068",
    "0.031",
    "0.070",
    "-0.035",
    "-0.003",
    "0.049",
    "-0.013",
    "-0.003",
    "0.051",
    "0.079",
    "0.038",
    "-0.012",
    "0.038",
    "0.042",
    "-0.013",
    "-0.079",
    "0.009",
    "0.013",
    "0.024",
    "0.010",
    "-0.011",
    "-0.082",
    "0.027",
    "-0.030",
    "-0.004",
    "0.013",
    "-0.035",
    "-0.020",
    "0.055",
    "-0.061",
    "-0.055",
    "0.013",
    "-0.019",
    "-0.013",
    "-0.016",
    "-0.069",
    "-0.045",
    "-0.049",
    "0.048",
    "-0.104",
    "-0.108",
    "0.000",
    "-0.000",
    "-0.086",
    "-0.088",
    "0.007",
    "-0.007",
    "-0.017",
    "0.045",
    "0.068",
    "-0.043",
    "0.086",
    "0.046",
    "0.068",
    "0.010",
    "-0.003",
    "0.059",
    "-0.036",
    "0.036",
    "0.066",
    "-0.038",
    "-0.062",
    "-0.045",
    "0.077",
    "0.043",
    "-0.021",
    "-0.022",
    "0.062",
    "-0.039",
    "0.028",
    "-0.013",
    "0.051",
    "-0.037",
    "0.055",
    "-0.066",
    "0.023",
    "0.005",
    "0.090",
    "0.005",
    "-0.083",
    "-0.055",
    "0.073",
    "-0.110",
    "-0.020",
    "0.113",
    "-0.053",
    "-0.058",
    "-0.024",
    "0.057",
    "0.127",
    "0.036",
    "-0.044",
    "0.017",
    "-0.025",
    "0.073",
    "0.043",
    "0.080",
    "-0.019",
    "-0.034",
    "0.097",
    "0.052",
    "0.011",
    "0.040",
    "0.002",
    "-0.008",
    "0.002",
    "0.014",
    "0.020",
    "-0.023",
    "0.021",
    "-0.009",
    "-0.051",
    "-0.016",
    "-0.090",
    "-0.061",
    "0.081",
    "0.002",
    "0.042",
    "0.044",
    "-0.025",
    "-0.095",
    "0.089",
    "-0.098",
    "-0.005",
    "0.035",
    "0.014",
    "-0.065",
    "-0.076",
    "0.012",
    "-0.050",
    "-0.056",
    "-0.057",
    "-0.018",
    "-0.002",
    "-0.002",
    "0.035",
    "-0.055",
    "-0.000",
    "-0.008",
    "0.027",
    "0.002",
    "0.010",
    "-0.022",
    "-0.022",
    "0.111",
    "0.005",
    "0.038",
    "0.004",
    "-0.067",
    "-0.028",
    "-0.044",
    "0.071",
    "0.019",
    "-0.049",
    "-0.104",
    "-0.044",
    "0.010",
    "0.042",
    "-0.014",
    "-0.034",
    "-0.025",
    "-0.014",
    "0.003",
    "0.033",
    "-0.022",
    "0.022",
    "0.071",
    "0.021",
    "0.024",
    "0.036",
    "-0.001",
    "-0.061",
    "-0.085",
    "0.007",
    "-0.039",
    "0.080",
    "-0.026",
    "-0.060",
    "0.061",
    "0.082",
    "-0.057",
    "0.005",
    "0.045",
    "0.024",
    "0.044",
    "0.091",
    "-0.051",
    "-0.011",
    "-0.069",
    "0.007",
    "0.073",
    "-0.043",
    "0.026",
    "-0.114",
    "-0.010",
    "0.022",
    "0.027",
    "0.004",
    "0.016",
    "0.004",
    "-0.021",
    "0.034",
  ],
  [
    "-0.010",
    "0.060",
    "0.025",
    "-0.006",
    "0.073",
    "0.017",
    "0.090",
    "-0.052",
    "0.100",
    "0.091",
    "0.007",
    "-0.020",
    "-0.075",
    "-0.017",
    "0.019",
    "0.041",
    "-0.011",
    "-0.059",
    "-0.125",
    "-0.049",
    "-0.044",
    "0.031",
    "0.038",
    "-0.032",
    "-0.091",
    "0.064",
    "-0.077",
    "0.085",
    "0.036",
    "-0.071",
    "0.021",
    "0.110",
    "0.033",
    "-0.026",
    "-0.062",
    "-0.027",
    "0.021",
    "-0.054",
    "0.057",
    "0.033",
    "-0.090",
    "-0.057",
    "0.027",
    "0.075",
    "-0.071",
    "-0.004",
    "0.054",
    "0.026",
    "0.078",
    "-0.049",
    "0.057",
    "-0.049",
    "-0.076",
    "0.007",
    "0.006",
    "0.037",
    "0.025",
    "0.022",
    "0.010",
    "-0.040",
    "-0.012",
    "-0.033",
    "-0.076",
    "0.071",
    "-0.021",
    "-0.063",
    "-0.001",
    "-0.055",
    "0.009",
    "-0.087",
    "0.021",
    "0.053",
    "-0.010",
    "-0.053",
    "-0.065",
    "0.042",
    "0.045",
    "0.033",
    "-0.020",
    "0.000",
    "0.038",
    "0.051",
    "0.005",
    "0.057",
    "0.046",
    "0.043",
    "-0.017",
    "0.012",
    "-0.057",
    "-0.035",
    "-0.019",
    "-0.007",
    "-0.074",
    "0.044",
    "-0.099",
    "0.041",
    "-0.024",
    "-0.120",
    "0.000",
    "-0.004",
    "0.056",
    "-0.000",
    "0.071",
    "-0.036",
    "-0.064",
    "-0.009",
    "-0.042",
    "0.078",
    "-0.024",
    "0.008",
    "-0.024",
    "-0.019",
    "-0.034",
    "0.078",
    "-0.061",
    "0.054",
    "0.034",
    "0.013",
    "-0.033",
    "0.030",
    "-0.017",
    "0.028",
    "-0.018",
    "0.067",
    "-0.070",
    "-0.018",
    "-0.037",
    "-0.000",
    "-0.006",
    "-0.032",
    "0.048",
    "-0.099",
    "0.094",
    "-0.052",
    "-0.037",
    "-0.026",
    "-0.037",
    "-0.030",
    "0.006",
    "-0.098",
    "-0.018",
    "-0.072",
    "0.039",
    "0.009",
    "-0.131",
    "0.021",
    "-0.076",
    "-0.010",
    "-0.017",
    "0.058",
    "0.022",
    "0.002",
    "0.009",
    "-0.031",
    "0.080",
    "0.031",
    "0.057",
    "-0.012",
    "0.116",
    "0.004",
    "0.018",
    "0.043",
    "0.029",
    "-0.007",
    "-0.025",
    "-0.026",
    "0.013",
    "0.122",
    "-0.004",
    "-0.045",
    "0.043",
    "-0.018",
    "0.062",
    "-0.039",
    "0.017",
    "0.008",
    "-0.026",
    "0.010",
    "0.023",
    "0.053",
    "0.069",
    "0.053",
    "-0.000",
    "0.010",
    "0.089",
    "-0.030",
    "-0.091",
    "-0.051",
    "-0.074",
    "0.086",
    "0.003",
    "0.011",
    "-0.008",
    "-0.011",
    "-0.029",
    "0.065",
    "0.005",
    "0.069",
    "0.043",
    "-0.019",
    "-0.017",
    "-0.037",
    "-0.006",
    "-0.087",
    "-0.009",
    "-0.013",
    "-0.005",
    "0.042",
    "-0.036",
    "0.017",
    "-0.087",
    "-0.051",
    "-0.150",
    "-0.002",
    "-0.026",
    "0.018",
    "0.022",
    "-0.015",
    "0.004",
    "-0.015",
    "-0.013",
    "-0.098",
    "-0.012",
    "0.000",
    "-0.023",
    "0.009",
    "-0.006",
    "0.066",
    "0.043",
    "-0.040",
    "0.027",
    "0.032",
    "-0.029",
    "0.046",
    "-0.056",
    "0.092",
    "-0.047",
    "0.080",
    "0.119",
    "0.049",
    "-0.067",
    "0.107",
    "0.009",
    "-0.051",
    "-0.068",
    "0.016",
    "-0.010",
    "0.044",
    "-0.030",
    "0.023",
    "-0.032",
    "-0.022",
    "-0.002",
    "-0.010",
    "0.003",
    "-0.037",
    "-0.112",
    "0.029",
    "0.030",
    "0.032",
    "-0.017",
    "-0.018",
    "-0.036",
    "0.084",
    "0.008",
    "-0.024",
    "0.159",
    "0.032",
    "-0.070",
    "-0.034",
    "-0.016",
    "0.080",
    "-0.062",
    "0.047",
    "0.045",
    "0.104",
    "0.030",
    "0.048",
    "0.097",
    "-0.016",
    "-0.064",
    "0.020",
    "-0.094",
    "-0.003",
    "0.020",
    "-0.027",
    "-0.079",
    "0.069",
    "0.108",
    "-0.033",
    "-0.064",
    "0.017",
    "0.019",
    "-0.023",
    "-0.019",
    "-0.019",
    "-0.064",
    "-0.032",
    "-0.003",
    "0.059",
    "0.060",
    "-0.045",
    "-0.067",
    "0.044",
    "-0.020",
    "-0.007",
    "0.047",
    "-0.047",
    "0.035",
    "-0.015",
    "-0.039",
    "0.047",
    "0.052",
    "0.001",
    "0.053",
    "0.025",
    "-0.029",
    "0.035",
    "0.013",
    "-0.000",
    "-0.034",
    "0.007",
    "-0.011",
    "0.043",
    "0.015",
    "-0.042",
    "0.106",
    "-0.013",
    "-0.002",
    "0.040",
    "-0.081",
    "0.034",
    "-0.020",
    "0.078",
    "0.060",
    "-0.016",
    "-0.113",
    "-0.043",
    "0.023",
    "0.074",
    "-0.061",
    "-0.046",
    "-0.088",
    "0.053",
    "-0.044",
    "0.030",
    "0.039",
    "0.017",
    "0.028",
    "0.078",
    "0.056",
    "0.053",
    "-0.011",
    "-0.030",
    "-0.085",
    "-0.016",
    "0.030",
    "0.069",
    "-0.062",
    "-0.122",
    "0.012",
    "0.073",
    "-0.018",
    "0.003",
    "0.014",
    "-0.038",
    "-0.053",
    "0.060",
    "-0.053",
    "0.018",
    "-0.104",
    "0.011",
    "0.043",
    "0.043",
    "-0.003",
    "-0.085",
    "-0.040",
    "-0.004",
    "0.013",
    "-0.008",
    "-0.029",
    "-0.036",
    "0.059",
    "-0.000",
  ],
]
`;
